commands/apps_dev.go

```diff
@@ -269,215 +265,81 @@ func RunAppsDevBuild(c *CmdConfig) error {
-	// 	charm.CheckmarkSuccess,
-	// 	charm.TextSuccess.S(res.Image),
-	// 	charm.TextSuccess.S(res.BuildDuration.Truncate(time.Second).String()),
-	// )
-
-	charm.TemplateBuffered(
-		charm.NewTextBox().Success(),
-		`{{ success checkmark }} Successfully built {{ success .img }} in {{ warning (duration .dur) }}`,
-		map[string]any{
-			"img": res.Image,
-			"dur": res.BuildDuration,
-		},
-	)
-	return nil
-}
-
-func newLogPager(ctx context.Context) *logPager {
-	// ctx, cancel := context.WithCancel(ctx)
-	return &logPager{
-		// ctx:    ctx,
-		// cancel: cancel,
-		start: time.Now(),
-	}
-}
-
-type logPager struct {
-	title string
-	// ctx    context.Context
-	// cancel context.CancelFunc
-	highPerf bool
-	p        *tea.Program
-	content  bytes.Buffer
-	ready    bool
-	viewport viewport.Model
-	start    time.Time
-}
-
-func (p *logPager) Write(b []byte) (int, error) {
-	n, err := p.content.Write(b)
-	p.p.Send(msgReload{})
-	return n, err
-}
-
-func (p *logPager) Close() error {
-	if p.p != nil {
-		p.p.Send(msgCancel{})
-	}
-	return nil
-}
-
-type msgCancel struct{}
-type msgReload struct{}
-type msgTick struct{}
-
-func (m *logPager) Init() tea.Cmd {
-	return m.timerTick()
-}
-
-func (m *logPager) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	charm.TemplatePrint(heredoc.Doc(`
+		{{success checkmark}} building {{lower (snakeToTitle .GetType)}} {{highlight .GetName}}{{nl}}{{nl}}`,
+	), componentSpec)
 	var (
-		cmd  tea.Cmd
-		cmds []tea.Cmd
+		wg        sync.WaitGroup
+		logWriter io.Writer
 	)
-
-	switch msg := msg.(type) {
-	case msgCancel:
-		return m, tea.Quit
-	case tea.KeyMsg:
-		if k := msg.String(); k == "ctrl+c" || k == "q" || k == "esc" {
-			return m, tea.Quit
-		}
-
-	case tea.WindowSizeMsg:
-		headerHeight := lipgloss.Height(m.headerView())
-		footerHeight := lipgloss.Height(m.footerView())
-		verticalMarginHeight := headerHeight + footerHeight
-
-		if !m.ready {
-			// Since this program is using the full size of the viewport we
-			// need to wait until we've received the window dimensions before
-			// we can initialize the viewport. The initial dimensions come in
-			// quickly, though asynchronously, which is why we wait for them
-			// here.
-			m.viewport = viewport.New(msg.Width, msg.Height-verticalMarginHeight)
-			m.viewport.YPosition = headerHeight
-			m.viewport.HighPerformanceRendering = m.highPerf
-			m.viewport.SetContent(m.content.String())
-			m.ready = true
-
-			// This is only necessary for high performance rendering, which in
-			// most cases you won't need.
-			//
-			// Render the viewport one line below the header.
-			m.viewport.YPosition = headerHeight + 1
-		} else {
-			m.viewport.Width = msg.Width
-			m.viewport.Height = msg.Height - verticalMarginHeight
+	if Interactive {
+		pager, err := charm.NewPager(
+			charm.PagerWithTitle("Building " + component),
+		)
+		if err != nil {
+			return fmt.Errorf("starting log pager: %w", err)
 		}
+		wg.Add(1)
+		go func() {
+			defer cancel()
```

Comment:
Why do we defer this cancel in multiple locations?

Reply:
Both the pager and the builder inherit the parent context:

```
	cli, err := c.Doit.GetContainerEngineClient()
	if err != nil {
		return err
	}

	ctx, cancel := context.WithCancel(ctx)
	defer cancel()
```

Once the builder exits we need to cancel the pager's context so that it can shut down cleanly. But we also want the builder to exit cleanly (remove containers etc) once the pager exits — this enables ctrl-c to work correctly while the pager is active. So we need to ensure that the context is canceled once either the pager or the builder exit, whichever comes first.